tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
     data = {
         'shortname': name
     }
-    response = client.post('/devices/add',
-                           data=json.dumps(data),
-                           content_type='application/json')
+    response = client.post('/devices/add', json=data)
     assert response.status_code == 201
-    data = json.loads(response.data)
+    data = response.json()
     assert 'id' in data
     assert 'shortname' in data
     assert 'name' in data
-    return data['id']
 
-
-def test_get_device(client):
-    id = test_add_device(client)
-
-    response = client.get('/devices/{}'.format(id))
-    assert response.status_code == 200
-    data = json.loads(response.data)
-    assert 'id' in data
-    assert 'shortname' in data
-    assert 'name' in data
+    client.delete('/devices/{}'.format(data['id']))
 
 
 def test_delete_device(client):
-    id = test_add_device(client)
+    id = add_device(client)
 
     response = client.delete('/devices/{}'.format(id))
     assert response.status_code == 204
 
 
-def test_get_dpi(client, dpi_id=0):
-    from ratbag_emu.device_handler import DeviceHandler
-
-    id = test_add_device(client)
+def test_get_device(client):
+    id = add_device(client)
 
-    response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
+    response = client.get('/devices/{}'.format(id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == DeviceHandler.devices[id].hprof.dpi[dpi_id if
-        dpi_id != 'active' else DeviceHandler.devices[id].hprof.active_dpi]
+    data = response.json()
+    assert 'id' in data
+    assert 'shortname' in data
+    assert 'name' in data
+    assert 'input_nodes' in data
 
+    client.delete('/devices/{}'.format(id))
 
-def test_set_dpi(client, dpi_id=0):
+
+def test_dpi(client, dpi_id=0):
     from ratbag_emu.device_handler import DeviceHandler
 
-    id = test_add_device(client)
+    id = add_device(client)
 
-    DeviceHandler.devices[id].hprof.dpi[dpi_id if dpi_id != 'active' else
-        DeviceHandler.devices[id].hprof.active_dpi] = 6666
+    data = 6666
+    response = client.put('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id),
+                          json=data)
 
     response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == 6666
-
+    assert response.json() == data
 
-def test_get_active_dpi(client):
-    test_get_dpi(client, 'active')
+    client.delete('/devices/{}'.format(id))
```

Comment:
you haven't tested the dpi here. You just tested that the state is somehow consistent.
I need the dpi to be validated by looking at the evdev events. (but maybe this will come later)